File: 2021/Day3/Day3.java
import java.io.*;
import java.util.*;

public class Day3 {
    public static void main(String[] args) throws IOException{
        Scanner scanner = new Scanner(new File("Day3.txt"));

        ArrayList<String> strings = new ArrayList<>();
        while (scanner.hasNext()) {
            strings.add(scanner.next());
        }

        int gamma = 0;
        for (int bit = 0; bit < strings.get(0).length(); bit++) {
            int n = 0;
            for (int string = 0; string < strings.size(); string++) {
                if (strings.get(string).charAt(strings.get(0).length() - 1 - bit) == '1') {
                    n++;
                }
            }

            if (n > strings.size() / 2) {
                gamma += Math.pow(2, bit);
            }
        }
        int epsilon = ((int) Math.pow(2, strings.get(0).length()) - 1) ^ gamma;

        System.out.println(gamma * epsilon);
    }
}

File: 2021/Day3/Day3.py
file = open("Day3.txt", "r").read().split()

bits = [0] * len(file[0])
for s in file:
    for i, bit in enumerate(s):
        bits[i] += int(bit)

gamma = int("".join(map(lambda x: "1" if x > len(file) / 2 else "0", bits)), 2)
epsilon = 2 ** len(bits) - 1 - gamma

print(gamma * epsilon)

strings = file
i = 0
while len(strings) > 1:
    strings = [string for string in strings if (sum(int(bits[i]) for bits in strings) >= len(strings) / 2) == (string[i] == "1")]
    i += 1

oxygen = strings[0]

strings = file
i = 0
while len(strings) > 1:
    strings = [string for string in strings if (sum(int(bits[i]) for bits in strings) < len(strings) / 2) == (string[i] == "1")]
    i += 1

co2 = strings[0]

print(int(oxygen, 2) * int(co2, 2))
